src/main/kotlin/pw/forst/tools/katlib/IterableExtensions.kt

```diff
@@ -497,3 +497,52 @@ inline fun <TItem> Iterable<TItem>.itemsToString(
     }
     return sb.toString().restrictLengthWithEllipsis(totalLength)
 }
+
+/**
+ * Transforms an Iterable into nullable elements while only keeping the non-null elements.
+ */
+fun <T, R> Iterable<T>.flatMapIndexedNotNull(function: (index: Int, T) -> Iterable<R>?): List<R> {
+    return flatMapIndexedTo(ArrayList(), function)
+}
+
+private inline fun <T, R, C : MutableCollection<in R>> Iterable<T>.flatMapIndexedTo(destination: C,
+                                                                                    transform: (index: Int, T) -> Iterable<R>?): C {
+
+    forEachIndexed { index, element ->
+        transform(index, element)?.let { elements ->
+            destination.addAll(elements)
+        }
+    }
+
+    return destination
+}
+
+/**
+ * Verifies the relationship between every element of an Iterable.
+ *
+ * Iterates through the elements invoking the validationFunction on each one.
+ * Returns false on the first element that does not pass the validation function, otherwise true.
+ *
+ * @param validationFunction is the accumulator function that verifies the elements.
+ */
+fun <S, T : S> Iterable<T>.validateElements(validationFunction: (acc: S, T) -> Boolean): Boolean {
```

Comment:
I'm not sure about name `validateElements` when this function provide possibility to validate relationship between ordered elements - so you're in a fact verifying that `n` and possibly `n+1` have some relationship

to me `validateElements` sounds like it provides a way how to validate collection of elements without any additional relationships

Reply:
Agreed. Will find something more fitting.

Reply:
I think `foldValidated` is the closest match I can think of. It's essentially a fold with early return, no?

Reply:
sounds gut to me